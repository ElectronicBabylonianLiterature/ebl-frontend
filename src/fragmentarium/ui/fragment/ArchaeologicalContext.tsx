import _ from 'lodash'
import React from 'react'
import { Form, Col } from 'react-bootstrap'

function ArchaeologicalContext({ archaeology }: { archaeology? }): JSX.Element {
  return (
    <Form>
      <Form.Row>
        <Form.Group as={Col} controlId={_.uniqueId('Archaeology-')}>
          <Form.Label>Excavation number</Form.Label>
          <Form.Control
            plaintext
            readOnly
            value={archaeology?.excavationNumber || 'N/A'}
          />
        </Form.Group>
      </Form.Row>
    </Form>
  )
}

export default ArchaeologicalContext
