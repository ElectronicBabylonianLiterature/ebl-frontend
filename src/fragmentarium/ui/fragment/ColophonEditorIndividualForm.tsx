import React, { useState } from 'react'
import { Accordion, Button, Form, Row } from 'react-bootstrap'
import {
  Colophon,
  IndividualAttestation,
  IndividualType,
} from 'fragmentarium/domain/Colophon'
import _ from 'lodash'
import ListForm from 'common/List'
import { BrokenAndUncertainSwitches } from 'common/BrokenAndUncertain'
import FragmentService from 'fragmentarium/application/FragmentService'
import { getSelectField } from './ColophonEditorIndividualInputs'
import Bluebird from 'bluebird'

export interface IndividualProps {
  fragmentService: FragmentService
  individual: IndividualAttestation
  index: number
  onChange: (individual: IndividualAttestation, index: number) => void
}

export const ColophonIndividualsInput = ({
  fragmentService,
  onChange,
  label = 'Individuals',
  collapsed = false,
  individualsProp = [],
}: {
  fragmentService: FragmentService
  onChange: (field: keyof Colophon, value) => void
  label?: string
  collapsed?: boolean
  individualsProp?: ReadonlyArray<IndividualAttestation>
}): JSX.Element => {
  const [individuals, setIndividuals] = useState([...individualsProp])
  const getFormFields = (individual, _onChange, index) => (
    <Accordion key={index}>
      <Accordion.Toggle as={Button} variant="link" eventKey={index.toString()}>
        {`Individual ${index + 1}. ${individual.toString()}`}
      </Accordion.Toggle>
      <Accordion.Collapse eventKey={index.toString()}>
        <IndividualForm
          index={index}
          fragmentService={fragmentService}
          onChange={(_individual: IndividualAttestation, index: number) => {
            const _individuals = [...individuals]
            _individuals[index] = _individual
            setIndividuals(_individuals)
            onChange('individuals', _individuals)
          }}
          individual={individual}
          key={index}
        />
      </Accordion.Collapse>
    </Accordion>
  )
  return (
    <ListForm
      value={individuals}
      onChange={(_individuals: IndividualAttestation[]) => {
        setIndividuals(_individuals)
        onChange('individuals', _individuals)
      }}
      label={label}
      noun="Individual"
      defaultValue={new IndividualAttestation({})}
      collapsed={collapsed}
    >
      {getFormFields}
    </ListForm>
  )
}

type IndividualFieldName = 'type' | 'name' | 'sonOf' | 'grandsonOf' | 'family'

const setBrokenOrUncertain = (
  checked: boolean,
  field: 'isBroken' | 'isUncertain',
  keyField: 'name' | 'family' | 'sonOf' | 'grandsonOf' | 'type',
  individual: IndividualAttestation
): IndividualAttestation => {
  if (keyField === 'type') {
    return individual.setTypeField({
      ...individual[keyField],
      [field]: checked,
    })
  } else {
    return individual.setNameField(keyField, {
      ...individual[keyField],
      [field]: checked,
    })
  }
}

const getIndividualField = ({
  individualProps,
  key,
}: {
  individualProps: IndividualProps
  key: string
}): JSX.Element => {
  const { individual, fragmentService, onChange, index } = individualProps
  const fieldProps = getValueAndOptionsByKey(
    key as IndividualFieldName,
    individual,
    fragmentService
  )
  const props = {
    onChange,
    key,
    isClearable: true,
    placeholder: _.startCase(key),
    ...fieldProps,
  }

  const getBrokenOrUncertainMethod = (field: 'isBroken' | 'isUncertain') => (
    checked: boolean
  ) => {
    const _individual = setBrokenOrUncertain(
      checked,
      field,
      key as 'name' | 'family' | 'sonOf' | 'grandsonOf' | 'type',
      individual
    )
    onChange(_individual, index)
  }

  return (
    <Form.Group key={`${key}-col`}>
      <Form.Label>{_.startCase(key)}</Form.Label>
      {getSelectField({ key, individualProps, props })}
      <Row key={`${key}-row`}>
        <BrokenAndUncertainSwitches
          key={`${key}-broken-uncertain`}
          name={key}
          id={index}
          setBroken={getBrokenOrUncertainMethod('isBroken')}
          setUncertain={getBrokenOrUncertainMethod('isUncertain')}
          isBroken={individual[key]?.isBroken}
          isUncertain={individual[key]?.isUncertain}
        />
      </Row>
    </Form.Group>
  )
}

const IndividualForm = (individualProps: IndividualProps): JSX.Element => {
  const individualFieldNames = [
    'name',
    'sonOf',
    'grandsonOf',
    'family',
    'nativeOf',
    'type',
  ]
  return (
    <>
      {individualFieldNames.map((key) =>
        getIndividualField({ individualProps, key })
      )}
    </>
  )
}

const getValueAndOptionsByKey = (
  key: IndividualFieldName,
  individual: IndividualAttestation,
  fragmentService: FragmentService
): {
  value?: { value: string; label: string }
  options?: readonly { value: string; label: string }[]
  loadOptions?
  handleChange?
} => {
  return {
    ...(individual[key]?.value && {
      value: {
        value: individual[key]?.value as string,
        label: individual[key]?.value as string,
      },
    }),
    ...(key === 'type' && {
      options: Object.values(IndividualType).map((type) => ({
        value: type,
        label: type,
      })),
    }),
    ...(key !== 'type' && {
      loadOptions: getLoadOptionsMethod(fragmentService),
    }),
  }
}

export const getLoadOptionsMethod = (fragmentService: FragmentService) => (
  inputValue: string,
  callback: (options: { value: string; label: string }[]) => void
): Bluebird<void> =>
  fragmentService.fetchColophonNames(inputValue).then((entries) => {
    const options = entries.map((value) => ({
      value,
      label: value,
    }))
    callback(options)
  })
