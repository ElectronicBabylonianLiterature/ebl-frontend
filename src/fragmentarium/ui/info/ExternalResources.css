.ExternalResources {
  text-align: center;
}

.ExternalResources a {
  display: block;
}

.ExternalResources__image {
  height: auto;
  max-height: 3.5em;
  max-width: 100%;
  margin: 0.2em;
}
