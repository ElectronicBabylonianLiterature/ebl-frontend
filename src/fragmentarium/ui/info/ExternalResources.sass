.ExternalResources
  &__image
    height: auto
    max-height: 3.5em
    max-width: 100%
    margin: 0.2em

  &__items
    padding: 0
    list-style-position: inside
