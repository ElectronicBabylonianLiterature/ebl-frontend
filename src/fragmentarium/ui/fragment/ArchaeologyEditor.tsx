import React, { ChangeEvent, Component, FormEvent } from 'react'
import _ from 'lodash'
import { Form, Col, Button } from 'react-bootstrap'
import Select, { ValueType } from 'react-select'
import {
  Archaeology,
  ArchaeologyDto,
  Findspot,
  SiteKey,
  excavationSites,
} from 'fragmentarium/domain/archaeology'
import { Fragment } from 'fragmentarium/domain/fragment'
import withData from 'http/withData'
import { FindspotService } from 'fragmentarium/application/FindspotService'

interface Props {
  archaeology?: Archaeology
  updateArchaeology: (archaeology: ArchaeologyDto) => Promise<Fragment>
  findspots: readonly Findspot[]
  disabled?: boolean
}

interface State {
  excavationNumber: string
  site: SiteKey
  isRegularExcavation: boolean
  findspotId: number | null
  findspot: Findspot | null
  error: Error | null
}

const excavationOptions = [
  {
    value: '',
    label: '-',
  },
  ..._.values(excavationSites).map((site) => ({
    value: site.name,
    label: site.name,
  })),
]

class ArchaeologyEditor extends Component<Props, State> {
  private isDirty = false
  private originalState: State
  private updateArchaeology: (archaeology: ArchaeologyDto) => Promise<Fragment>
  private findspots: ReadonlyMap<number, Findspot>
  private findspotOptions

  constructor(props: Props) {
    super(props)
    const archaeology = props.archaeology || {}

    this.state = {
      excavationNumber: archaeology.excavationNumber || '',
      site: (archaeology.site?.name || '') as SiteKey,
      isRegularExcavation: archaeology.isRegularExcavation ?? true,
      error: null,
      findspotId: archaeology.findspotId || null,
      findspot: archaeology.findspot || null,
    }
    this.originalState = { ...this.state }
    this.updateArchaeology = props.updateArchaeology

    this.findspots = new Map(
      props.findspots.map((findspot) => [findspot.id, findspot])
    )
    this.findspotOptions = props.findspots.map((findspot) => ({
      value: findspot.id,
      label: findspot.toString(),
    }))
  }

  updateState = (property: string) => (
    value: string | boolean | number | Findspot | null
  ): void => {
    const updatedState = {
      ...this.state,
      [property]: value,
    }
    this.isDirty = !_.isEqual(this.originalState, updatedState)
    this.setState(updatedState)
  }
  updateFindspotState = (
    findspotId: number | null,
    findspot: Findspot | null
  ): void => {
    this.updateState('findspotId')(findspotId)
    this.updateState('findspot')(findspot)
  }

  updateExcavationNumber = (event: ChangeEvent<HTMLInputElement>): void =>
    this.updateState('excavationNumber')(event.target.value)

  updateSite = (
    event: ValueType<typeof excavationOptions[number], false>
  ): void => this.updateState('site')(event?.value || '')

  updateIsRegularExcavation = (event: ChangeEvent<HTMLInputElement>): void =>
    this.updateState('isRegularExcavation')(event.target.checked)

  updateFindspot = (
    event: ValueType<typeof this.findspotOptions[number], false>
  ): void => {
    if (_.isNull(event)) {
      this.updateFindspotState(null, null)
    } else {
      this.updateFindspotState(
        event.value,
        this.findspots.get(event.value) || null
      )
    }
  }

  submit = (event: FormEvent<HTMLElement>): void => {
    event.preventDefault()

    this.updateArchaeology({
      ..._.omitBy(
        { ...this.state, error: null },
        (value) => _.isNil(value) || value === ''
      ),
    })
      .then((_updatedFragment) => {
        this.originalState = { ...this.state }
        this.isDirty = false
        this.setState(this.originalState)
      })
      .catch((error) =>
        this.setState({
          ...this.state,
          error: error,
        })
      )
  }

  renderExcavationNumberForm = (): JSX.Element => (
    <Form.Group as={Col} controlId={_.uniqueId('excavationNumber-')}>
      <Form.Label>Excavation number</Form.Label>
      <Form.Control
        type="text"
        value={this.state.excavationNumber}
        onChange={this.updateExcavationNumber}
      />
    </Form.Group>
  )
  renderExcavationSiteForm = (): JSX.Element => (
    <Form.Group as={Col} controlId={_.uniqueId('excavationSite-')}>
      <Form.Label>Excavation site</Form.Label>
      <Select
        aria-label="select-site"
        options={excavationOptions}
        value={{
          value: this.state.site,
          label: this.state.site,
        }}
        onChange={this.updateSite}
        isSearchable={true}
        isClearable
      />
    </Form.Group>
  )
  renderIsRegularExcavationForm = (): JSX.Element => (
    <Form.Group>
      <Form.Check
        type="checkbox"
        id={_.uniqueId('isRegularExcavation-')}
        label="Regular Excavation"
        checked={this.state.isRegularExcavation}
        onChange={this.updateIsRegularExcavation}
      />
    </Form.Group>
  )
  renderFindspotForm = (): JSX.Element => (
    <Form.Group as={Col} controlId={_.uniqueId('findspot-')}>
      <Form.Label>Findspot</Form.Label>
      <Select
        aria-label="select-findspot"
        options={this.findspotOptions}
        value={{
          value: this.state.findspotId,
          label: this.state.findspot?.toString(),
        }}
        onChange={this.updateFindspot}
        isSearchable={true}
        isClearable
      />
    </Form.Group>
  )

  render(): JSX.Element {
    return (
      <Form onSubmit={this.submit} data-testid="archaeology-form">
        <Form.Row>{this.renderExcavationNumberForm()}</Form.Row>
        <Form.Row>{this.renderExcavationSiteForm()}</Form.Row>
        <Form.Row>{this.renderIsRegularExcavationForm()}</Form.Row>
        <Form.Row>{this.renderFindspotForm()}</Form.Row>
        <Button
          variant="primary"
          type="submit"
          disabled={this.props.disabled || !this.isDirty}
        >
          Save
        </Button>
      </Form>
    )
  }
}

export default withData<
  {
    archaeology?: Archaeology
    updateArchaeology: (archaeology: ArchaeologyDto) => Promise<Fragment>
    disabled?: boolean
  },
  { findspotService: FindspotService },
  readonly Findspot[]
>(
  ({ archaeology, updateArchaeology, disabled, data: findspots }) => {
    return (
      <ArchaeologyEditor
        archaeology={archaeology}
        updateArchaeology={updateArchaeology}
        findspots={findspots}
        disabled={disabled}
      />
    )
  },
  (props) => props.findspotService.fetchFindspots()
)
