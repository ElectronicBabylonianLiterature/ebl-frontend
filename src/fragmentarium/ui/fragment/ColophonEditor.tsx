import React, { useState } from 'react'
import { Form, Button, Row } from 'react-bootstrap'
import { Fragment } from 'fragmentarium/domain/fragment'
import {
  ColophonOwnershipInput,
  ColophonStatusInput,
  ColophonTypeInput,
  ColophonNotesToScribalProcessInput,
  ProvenanceAttestationInput,
} from './ColophonEditorInputs'
import FragmentService from 'fragmentarium/application/FragmentService'
import { ColophonIndividualsInput } from './ColophonEditorIndividualForm'
import { Colophon } from 'fragmentarium/domain/Colophon'

interface Props {
  fragment: Fragment
  updateColophon: (colophon: Colophon) => Promise<void>
  disabled?: boolean
  fragmentService: FragmentService
}

const ColophonEditor: React.FC<Props> = ({
  fragment,
  disabled,
  updateColophon,
  fragmentService,
}) => {
  // ToDo:
  // 3. Ensure that the form loads and dumps data correctly
  // 4. Side panel
  // 5. Tests

  const { colophon } = fragment
  const [formData, setFormData] = useState<Colophon>({
    colophonStatus: colophon?.colophonStatus,
    colophonOwnership: colophon?.colophonOwnership,
    colophonTypes: colophon?.colophonTypes,
    originalFrom: colophon?.originalFrom,
    writtenIn: colophon?.writtenIn,
    notesToScribalProcess: colophon?.notesToScribalProcess,
    individuals: colophon?.individuals,
  })
  const [error, setError] = useState<Error | null>(null)

  const handleChange = (field: keyof Colophon, value) => {
    setFormData((prevState) => ({
      ...prevState,
      [field]: value,
    }))
  }

  const submit = async (): Promise<void> => {
    try {
      await updateColophon(formData)
    } catch (error) {
      setError(error as Error)
    }
  }

  return (
    <>
      <Form>
        <Row>
          <ColophonStatusInput
            colophonStatus={formData.colophonStatus}
            onChange={handleChange}
          />
        </Row>
        <Row>
          <ColophonTypeInput
            colophonTypes={formData.colophonTypes}
            onChange={handleChange}
          />
        </Row>
        <Row>
          <ColophonOwnershipInput
            colophonOwnership={formData.colophonOwnership}
            onChange={handleChange}
          />
        </Row>
        <Row>
          <ProvenanceAttestationInput
            {...{
              onChange: handleChange,
              fragmentService,
              fieldName: 'originalFrom',
              colophon: formData,
            }}
          />
        </Row>
        <Row>
          <ProvenanceAttestationInput
            {...{
              onChange: handleChange,
              fragmentService,
              fieldName: 'writtenIn',
              colophon: formData,
            }}
          />
        </Row>
        <ColophonIndividualsInput
          individualsProp={formData.individuals}
          onChange={handleChange}
          fragmentService={fragmentService}
        />
        <Row>
          <ColophonNotesToScribalProcessInput
            notesToScribalProcess={formData.notesToScribalProcess}
            onChange={handleChange}
          />
        </Row>
      </Form>
      <Button
        variant="primary"
        type="submit"
        aria-label="save-colophon"
        disabled={disabled || error !== null}
        onClick={(event) => {
          event.preventDefault()
          submit()
        }}
      >
        Save
      </Button>
    </>
  )
}

export default ColophonEditor
