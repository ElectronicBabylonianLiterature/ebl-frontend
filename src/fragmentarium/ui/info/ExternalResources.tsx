import React from 'react'
import { Fragment } from 'fragmentarium/domain/fragment'
import './ExternalResources.sass'
import {
  BdtnsLink,
  CdliLink,
  ArchibabLink,
  UrOnlineLink,
  HilprechtJenaLink,
  HilprechtHeidelbergLink,
  AchemenetLink,
  BmIdLink,
  YalePeabodyLink,
  OraccLinks,
} from './ExternalNumberLink'
import _ from 'lodash'

type ExternalLinkComponent = ({ number }: { number: string }) => JSX.Element

export default function ExternalResources({
  fragment,
}: {
  readonly fragment: Fragment
}): JSX.Element {
  const externalNumbers: ReadonlyArray<[string, ExternalLinkComponent]> = [
    [fragment.bmIdNumber, BmIdLink],
    [fragment.cdliNumber, CdliLink],
    [fragment.bdtnsNumber, BdtnsLink],
    [fragment.archibabNumber, ArchibabLink],
    [fragment.urOnlineNumber, UrOnlineLink],
    [fragment.hilprechtJenaNumber, HilprechtJenaLink],
    [fragment.hilprechtHeidelbergNumber, HilprechtHeidelbergLink],
    [fragment.achemenetNumber, AchemenetLink],
    [fragment.yalePeabodyNumber, YalePeabodyLink],
  ]
  return (
    <ul className="ExternalResources__items">
      {externalNumbers.map(
        ([number, LinkComponent], index) =>
          number && (
            <li key={index}>
              <LinkComponent number={number} />
            </li>
          )
      )}
      {!_.isEmpty(fragment.oraccNumbers) && (
        <li>
          <OraccLinks
            projects={fragment.oraccNumbers}
            cdliNumber={fragment.cdliNumber}
          />
        </li>
      )}
    </ul>
  )
}
