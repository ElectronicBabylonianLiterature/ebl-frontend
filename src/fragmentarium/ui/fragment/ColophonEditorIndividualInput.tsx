import React from 'react'
import { Form, Row } from 'react-bootstrap'
import { IndividualAttestation, IndividualType } from './ColophonEditor'
import _ from 'lodash'
import ListForm from 'common/List'
import { BrokenAndUncertainSwitches } from 'common/BrokenAndUncertain'
import Select from 'react-select'
import CreatableSelect from 'react-select/creatable'
import FragmentService from 'fragmentarium/application/FragmentService'
import ProvenanceSearchForm from '../ProvenanceSearchForm'
import { Provenances } from 'corpus/domain/provenance'

export const ColophonIndividualsInput = ({
  fragmentService,
  onChange,
  label = 'Individuals',
  collapsed = false,
  individuals = [new IndividualAttestation()],
}: {
  fragmentService: FragmentService
  onChange
  label?: string
  collapsed?: boolean
  individuals?: ReadonlyArray<IndividualAttestation>
}): JSX.Element => {
  const _individuals =
    individuals.length > 0 ? individuals : [new IndividualAttestation()]
  return (
    <ListForm
      value={_individuals}
      onChange={onChange}
      label={label}
      noun="Individual"
      defaultValue={_individuals}
      collapsed={collapsed}
    >
      {() =>
        _individuals.map((individual, index) => {
          return (
            <IndividualForm
              fragmentService={fragmentService}
              onChange={onChange}
              individual={individual}
              key={index}
            />
          )
        })
      }
    </ListForm>
  )
}

type IndividualFieldName = 'type' | 'name' | 'sonOf' | 'grandsonOf' | 'family'

const getValueAndOptionsByKey = async (
  key: IndividualFieldName,
  individual: IndividualAttestation,
  fragmentService: FragmentService
): Promise<{
  options: readonly { value: string; label: string }[]
  value: { value: string; label: string }
}> =>
  key === 'type'
    ? {
        value: { value: individual[key] ?? '', label: individual[key] ?? '' },
        options: Object.values(IndividualType).map((type) => ({
          value: type ?? '',
          label: type ?? '',
        })),
      }
    : {
        value: {
          value: individual[key]?.value ?? '',
          label: individual[key]?.value ?? '',
        },
        options: await fragmentService
          .fetchColophonNames('ToDo: add query here')
          .then((names) => names.map((name) => ({ value: name, label: name }))),
      }

const IndividualForm = ({
  fragmentService,
  onChange,
  individual,
}: {
  fragmentService: FragmentService
  onChange
  individual: IndividualAttestation
}): JSX.Element => {
  const nameFields = ['name', 'sonOf', 'grandsonOf', 'family']
  const getSelectField = (props, key: string) => {
    return key === 'nativeOf' ? (
      <ProvenanceSearchForm
        fragmentService={fragmentService}
        onChange={(value) => {
          individual.setNativeOf({
            ...individual.nativeOf,
            value: value && Provenances[value] ? Provenances[value] : null,
          })
        }}
        value={individual?.nativeOf?.value?.name ?? null}
        placeholder="Native Of"
      />
    ) : nameFields.includes(key) ? (
      <CreatableSelect {...props} />
    ) : (
      <Select {...props} />
    )
  }
  const setBrokenOrUncertain = (
    checked: boolean,
    field: 'broken' | 'uncertain',
    keyField: 'name' | 'family' | 'sonOf' | 'grandsonOf',
    individual: IndividualAttestation
  ) => {
    individual.setNameField(keyField, {
      ...individual[keyField],
      [field]: checked,
    })
    console.log(keyField, field, individual)
  }
  const individualFields = [...nameFields, 'nativeOf', 'type'].map((key) => {
    const props = {
      onChange,
      key,
      isClearable: true,
      placeholder: _.startCase(key),
      ...getValueAndOptionsByKey(
        key as IndividualFieldName,
        individual,
        fragmentService
      ),
    }

    const getBrokenOrUncertainMethod = (field: 'broken' | 'uncertain') => (
      checked: boolean
    ) =>
      setBrokenOrUncertain(
        checked,
        field,
        key as 'name' | 'family' | 'sonOf' | 'grandsonOf',
        individual
      )

    return (
      <Form.Group key={`${key}-col`}>
        <Form.Label>{_.startCase(key)}</Form.Label>
        {getSelectField(props, key)}
        {key !== 'type' && (
          <Row key={`${key}-row`}>
            <BrokenAndUncertainSwitches
              key={`${key}-broken-uncertain`}
              name={key}
              setBroken={getBrokenOrUncertainMethod('broken')}
              setUncertain={getBrokenOrUncertainMethod('uncertain')}
            />
          </Row>
        )}
      </Form.Group>
    )
  })
  return <>{individualFields}</>
}
