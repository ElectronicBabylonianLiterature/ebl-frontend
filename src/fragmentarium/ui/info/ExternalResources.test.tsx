import React from 'react'
import { render, screen } from '@testing-library/react'
import ExternalResources from './ExternalResources'

import {
  externalNumbersFactory,
  fragmentFactory,
} from 'test-support/fragment-fixtures'
import { Fragment } from 'fragmentarium/domain/fragment'

const cdliNumber = 'P 0000+q'
const bmIdNumber = 'bm 00000+q'
const bdtnsNumber = 'bdtns123'
const archibabNumber = 'A38'
const urOnlineNumber = 'U5'
const hilprechtJenaNumber = 'H42'
const hilprechtHeidelbergNumber = 'H007'
const yalePeabodyNumber = 'y123'
let fragment: Fragment
let container: HTMLElement

describe('external resources', () => {
  beforeEach(async () => {
    fragment = fragmentFactory.build(
      {},
      {
        associations: {
          externalNumbers: externalNumbersFactory.build({
            cdliNumber,
            bmIdNumber,
            bdtnsNumber,
            archibabNumber,
            urOnlineNumber,
            hilprechtJenaNumber,
            hilprechtHeidelbergNumber,
            yalePeabodyNumber,
          }),
        },
      }
    )
    container = render(<ExternalResources fragment={fragment} />).container
  })

  test.each([
    ['CDLI', 'https://cdli.mpiwg-berlin.mpg.de/', cdliNumber],
    [
      'The British Museum',
      'https://www.britishmuseum.org/collection/object/',
      bmIdNumber,
    ],
    ['BDTNS', 'http://bdtns.filol.csic.es/', bdtnsNumber],
    ['Archibab', 'http://www.archibab.fr/', archibabNumber],
    ['Ur Online', 'http://www.ur-online.org/subject/', urOnlineNumber],
    [
      'Hilprecht Collection',
      'https://hilprecht.mpiwg-berlin.mpg.de/object3d/',
      hilprechtJenaNumber,
    ],
    [
      'Hilprecht Collection – HeiCuBeDa',
      'https://doi.org/10.11588/heidicon/',
      hilprechtHeidelbergNumber,
    ],
    [
      'Yale Babylonian Collection',
      'https://collections.peabody.yale.edu/search/Record/YPM-',
      yalePeabodyNumber,
    ],
  ])('%s number is shown', async (label, link, number) => {
    expect(screen.getByLabelText(`${label} text ${number}`)).toHaveAttribute(
      'href',
      `${link}${encodeURIComponent(number)}`
    )
  })
  test('Snapshot', async () => expect(container).toMatchSnapshot())
})

describe('missing external resources', () => {
  beforeEach(async () => {
    fragment = fragmentFactory.build(
      {},
      {
        associations: {
          externalNumbers: {},
        },
      }
    )
    render(<ExternalResources fragment={fragment} />)
  })

  test.each([
    'CDLI',
    'The British Museum',
    'BDTNS',
    'Archibab',
    'Ur Online',
    'Hilprecht Collection',
    'Hilprecht Collection – HeiCuBeDa',
    'Yale Babylonian Collection',
  ])('Mising %s number is not shown', async (label) => {
    expect(screen.queryByText(label)).not.toBeInTheDocument()
  })
})
