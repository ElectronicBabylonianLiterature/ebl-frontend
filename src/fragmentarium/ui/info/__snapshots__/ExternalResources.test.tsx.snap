// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`external resources Snapshot 1`] = `
<div>
  <ul
    class="ExternalResources__items"
  >
    <li>
      The British Museum (
      <a
        aria-label="The British Museum text bm 00000+q"
        href="https://www.britishmuseum.org/collection/object/bm%2000000%2Bq"
        rel="noopener noreferrer"
        target="_blank"
      >
        bm 00000+q
      </a>
      )
    </li>
    <li>
      CDLI (
      <a
        aria-label="CDLI text P 0000+q"
        href="https://cdli.mpiwg-berlin.mpg.de/P%200000%2Bq"
        rel="noopener noreferrer"
        target="_blank"
      >
        P 0000+q
      </a>
      )
    </li>
    <li>
      BDTNS (
      <a
        aria-label="BDTNS text bdtns123"
        href="http://bdtns.filol.csic.es/bdtns123"
        rel="noopener noreferrer"
        target="_blank"
      >
        bdtns123
      </a>
      )
    </li>
    <li>
      Archibab (
      <a
        aria-label="Archibab text A38"
        href="http://www.archibab.fr/A38"
        rel="noopener noreferrer"
        target="_blank"
      >
        A38
      </a>
      )
    </li>
    <li>
      Ur Online (
      <a
        aria-label="Ur Online text U5"
        href="http://www.ur-online.org/subject/U5"
        rel="noopener noreferrer"
        target="_blank"
      >
        U5
      </a>
      )
    </li>
    <li>
      Hilprecht Collection (
      <a
        aria-label="Hilprecht Collection text H42"
        href="https://hilprecht.mpiwg-berlin.mpg.de/object3d/H42"
        rel="noopener noreferrer"
        target="_blank"
      >
        H42
      </a>
      )
    </li>
    <li>
      Hilprecht Collection – HeiCuBeDa (
      <a
        aria-label="Hilprecht Collection – HeiCuBeDa text H007"
        href="https://doi.org/10.11588/heidicon/H007"
        rel="noopener noreferrer"
        target="_blank"
      >
        H007
      </a>
      )
    </li>
    <li>
      Achemenet (
      <a
        aria-label="Achemenet text H00744"
        href="http://www.achemenet.com/en/item/?/textual-sources/texts-by-languages-and-scripts/babylonian/H00744"
        rel="noopener noreferrer"
        target="_blank"
      >
        H00744
      </a>
      )
    </li>
    <li>
      NaBuCCo (
      <a
        aria-label="NaBuCCo text H00765"
        href="https://nabucco.acdh.oeaw.ac.at/archiv/tablet/detail/H00765"
        rel="noopener noreferrer"
        target="_blank"
      >
        H00765
      </a>
      )
    </li>
    <li>
      Yale Babylonian Collection (
      <a
        aria-label="Yale Babylonian Collection text y123"
        href="https://collections.peabody.yale.edu/search/Record/YPM-y123"
        rel="noopener noreferrer"
        target="_blank"
      >
        y123
      </a>
      )
    </li>
  </ul>
</div>
`;
